src/audio/component.c

```diff
@@ -328,162 +328,273 @@ int comp_verify_params(struct comp_dev *dev, uint32_t flag,
+	if (blob_handler->data_new && cdata->msg_index == 0) {
+		comp_err(blob_handler->dev, "comp_data_blob_set_cmd(), busy with previous request");
+		return -EBUSY;
+	}
+
 	/* in case when the current package is the first, we should allocate
 	 * memory for whole model data
 	 */
 	if (!cdata->msg_index) {
-		ret = comp_alloc_model_data(dev, model, cdata->data->size);
 		/* in case when required model size is equal to zero we do not
 		 * allocate memory and should just return 0
 		 */
-		if (ret < 0 || !cdata->data->size)
-			return ret;
+		if (!cdata->data->size)
+			return 0;
+
+		blob_handler->data_new = rballoc(0, SOF_MEM_CAPS_RAM,
+						 cdata->data->size);
+		if (!blob_handler->data_new) {
+			comp_err(blob_handler->dev, "comp_data_blob_set_cmd(): blob_handler->data_new allocation failed.");
+			return -ENOMEM;
+		}
+
+		blob_handler->data_size = cdata->data->size;
+		blob_handler->data_ready = false;
+		blob_handler->data_pos = 0;
 	}
 
 	/* return an error in case when we do not have allocated memory for
 	 * model data
 	 */
-	if (!model->data) {
-		comp_err(dev, "comp_set_model(): buffer not allocated");
+	if (!blob_handler->data_new) {
+		comp_err(blob_handler->dev, "comp_data_blob_set_cmd(): buffer not allocated");
 		return -ENOMEM;
 	}
 
-	if (!cdata->elems_remaining) {
-		/* when we receive the last package and do not fill the whole
-		 * allocated buffer, we return an error
-		 */
-		if (cdata->num_elems + model->data_pos < model->data_size) {
-			comp_err(dev, "comp_set_model(): not enough data to fill the buffer");
-			// TODO: handle this situation
-
-			return -EINVAL;
-		}
+	ret = memcpy_s((char *)blob_handler->data_new + blob_handler->data_pos,
+		       blob_handler->data_size - blob_handler->data_pos,
+		       cdata->data->data, cdata->num_elems);
+	assert(!ret);
 
-		/* the whole data were received properly */
-		done = true;
-		comp_dbg(dev, "comp_set_model(): final package received");
-	}
+	blob_handler->data_pos += cdata->num_elems;
 
-	/* return an error in case when received data exceed allocated
-	 * memory
-	 */
-	if (cdata->num_elems > model->data_size - model->data_pos) {
-		comp_err(dev, "comp_set_model(): too much data");
-		return -EINVAL;
-	}
+	if (!cdata->elems_remaining) {
+		comp_dbg(blob_handler->dev, "comp_data_blob_set_cmd(): final package received");
 
-	ret = memcpy_s((char *)model->data + model->data_pos,
-		       model->data_size - model->data_pos,
-		       cdata->data->data, cdata->num_elems);
-	assert(!ret);
+		/* The new configuration is OK to be applied */
+		blob_handler->data_ready = true;
 
-	/* update data_pos variable with received number of elements (num_elem)
-	 */
-	model->data_pos += cdata->num_elems;
+		/* If component state is READY we can omit old
+		 * configuration immediately. When in playback/capture
+		 * the new configuration presence is checked in copy().
+		 */
+		if (blob_handler->dev->state ==  COMP_STATE_READY) {
```

Comment:
Can we have a ```comp_data_blob_done()``` to free that data that wil be called by client when it's done with the new blob. This will avoid any state change races.

Reply:
@lgirdwood 
I do not think there will appear some state change races. This code is very largely based on previous equalizer blob code, and as far as I know there were no such errors. Above condition ```if (blob_handler->dev->state ==  COMP_STATE_READY) {``` is only to force update config in READY state, because during streaming (playback/capture) component needs to be reconfigured. What do you think?

Reply:
The mixer and stream updates are asynchronous in userspace and the kernel. The FW xrun handling is also asynchronous so it is best for the component to free the data when it's done with it.

Reply:
@lgirdwood 
I have one concern. As I understand correctly, if we free config just after we use it, we will make it impossible for driver to fetch config using get_model() later.

Reply:
IIRC, the driver will cache these - @ranj063 @bardliao please confirm.

Reply:
I think even if it doesn't cache it currently it should be added in at some point.

Reply:
> IIRC, the driver will cache these - @ranj063 @bardliao please confirm.

Yes, kcontrol values are cached. Currently, we don't even get the data from FW side via ipc when a kcontrol_get() function is invoked.

Reply:
@bkokoszx is this resolved in the latest push ? (it was pushed before @bardliao confirmed).

Reply:
@lgirdwood 
I did not resolve it yet. I still have on question. In that case, why do we have get_cmd in firmware (assuming that kcontrols value are cached)?

Reply:
@bkokoszx we have get() for "volatile"  kcontrols and getting initial values (e.g after FW or topology load)

Reply:
@lgirdwood 
I have one more concern (I hope the last one :)). I've added ``comp_init_data_blob()`` function (it is called e.g. in ``eq_fir_new()``). It allocates data for config and initializes it with some values. If we invoke ``comp_data_blob_done()`` just after that, we won't be able to fetch this "initial values" by get_cmd.

Reply:
@bkokoszx I would expect the blob to be allocated/freed in the get()/set() flow. But IIRC some components will also install a default config at new(). What's the usage flow here for this update ? My expectattion is that we only only allocate the memory when needed ?

Reply:
@lgirdwood 
As you wrote, even before this PR, we have situations that component install and allocate default config in ``new()`` (e.g. eq_fir, eq_iir). The flow (for eq_fir/irr) was:
1. Allocate new config in new() if it is given.
2. Configure component in prepare() with allocated config (config allocated in ``new()`` or config sent by ``set()`` function).
3. Reconfigure component with new config if available (in copy())

My change (this PR) only wraps config init allocation into ``comp_init_blob_data()``

Reply:
So where do you free(). In theory you could free at then end of 2 and also at the end of 3 ?

Reply:
@lgirdwood 
I'm doing complete ``free()``, as it was earlier, in ``component_free()`` function. Of course, if we get new config in ``copy()``, we overwrite old with new one (and also free the old one):
``` c
void *comp_get_data_blob(struct comp_data_blob_handler *blob_handler,
			 size_t *size, uint32_t *crc)
{
	assert(blob_handler);

	comp_dbg(blob_handler->dev, "comp_get_data_blob()");

	if (size)
		*size = 0;

	/* Function returns new data blob if available */
	if (comp_is_new_data_blob_available(blob_handler)) {
		comp_dbg(blob_handler->dev, "comp_get_data_blob(): new data available");

		/* Free "old" data blob and set data to data_new pointer */
		rfree(blob_handler->data);
		blob_handler->data = blob_handler->data_new;
		blob_handler->data_new = NULL;
		blob_handler->data_ready = false;
	}
```

Reply:
ok, got you - btw does this mean we keep 2 copies of a config ? i.e. the blob and the blob contents stored in a component coefficient buffer ?

Reply:
@lgirdwood 
We store one copy of config (I do not understand what you mean by "component coefficient buffer"). We allocate memory for blob_handler->data (and for blob_handler->data_new if new config comes). ``comp_get_data_blob`` only returns pointer to allocated memory, which we can cast to specific component config struct. E.g. in ``eq_fir_prepare()``:
```
cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
```